controllers/processingjob/processingjob_controller.go

```diff
@@ -0,0 +1,306 @@
+		return r.updateStatusAndReturnError(ctx, "", unknownStateError)
+	}
+
+	status := string(ctx.ProcessingJobDescription.ProcessingJobStatus)
+	additional := controllers.GetOrDefault(ctx.ProcessingJobDescription.FailureReason, "")
+
+	if err = r.updateStatusWithAdditional(ctx, status, additional); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Initialize fields on the context object which will be used later.
+func (r *Reconciler) initializeContext(ctx *reconcileRequestContext) error {
+
+	ctx.ProcessingJobName = controllers.GetGeneratedJobName(ctx.ProcessingJob.ObjectMeta.GetUID(), ctx.ProcessingJob.ObjectMeta.GetName(), MaxProcessingJobNameLength)
+
+	ctx.Log.Info("ProcessingJob", "Using job name: ", ctx.ProcessingJobName)
+
+	awsConfig, err := r.awsConfigLoader.LoadAwsConfigWithOverrides(*ctx.ProcessingJob.Spec.Region, ctx.ProcessingJob.Spec.SageMakerEndpoint)
+	if err != nil {
+		ctx.Log.Error(err, "Error loading AWS config")
+		return err
+	}
+
+	ctx.SageMakerClient = r.createSageMakerClient(awsConfig)
+	ctx.Log.Info("Loaded AWS config")
+
+	return nil
+}
+
+// Creates the training job in SageMaker
+func (r *Reconciler) createProcessingJob(ctx reconcileRequestContext) error {
+	var createProcessingJobInput sagemaker.CreateProcessingJobInput
+
+	createProcessingJobInput = sdkutil.CreateCreateProcessingJobInputFromSpec(ctx.ProcessingJob.Spec, &ctx.ProcessingJobName)
+
+	ctx.Log.Info("Creating ProcessingJob in SageMaker", "input", createProcessingJobInput)
+
+	if _, err := ctx.SageMakerClient.CreateProcessingJob(ctx, &createProcessingJobInput); err != nil {
+		return errors.Wrap(err, "Unable to create Processing Job")
+	}
+
+	return nil
+}
+
+// Removes the finalizer held by our controller.
+func (r *Reconciler) removeFinalizer(ctx reconcileRequestContext) error {
+	var err error
+
+	ctx.ProcessingJob.ObjectMeta.Finalizers = controllers.RemoveString(ctx.ProcessingJob.ObjectMeta.Finalizers, controllers.SageMakerResourceFinalizerName)
+	if err = r.Update(ctx, ctx.ProcessingJob); err != nil {
+		return errors.Wrap(err, "Failed to remove finalizer")
+	}
+	ctx.Log.Info("Finalizer has been removed")
+
+	return nil
+}
+
+// If this function returns an error, the status update has failed, and the reconciler should always requeue.
+// This prevents the case where a terminal status fails to persist to the Kubernetes datastore yet we stop
+// reconciling and thus leave the job in an unfinished state.
+func (r *Reconciler) updateStatus(ctx reconcileRequestContext, processingJobStatus string) error {
+	return r.updateStatusWithAdditional(ctx, processingJobStatus, "")
+}
+
+func (r *Reconciler) updateStatusAndReturnError(ctx reconcileRequestContext, processingJobStatus string, reconcileErr error) error {
+	if len(processingJobStatus) == 0 {
+		if clientwrapper.IsRecoverableError(reconcileErr) {
+			processingJobStatus = ReconcilingProcessingJobStatus
+		} else {
+			processingJobStatus = controllers.ErrorStatus
```

Comment:
As mentioned in the bug bash,  we discussed having an `Error` and `Reconciling` status. Why is there an additonal `failed` status too ? Is this the case across operators ?

Reply:
yes, this is new and will be clarified in documentation

`Failed` status denotes the job was created on SageMaker and failed. The operator wont attempt to reconcile once the job has failed

`Error` status can be related to problems with initialization of operator, cluster/IAM role setup etc. basically when controller cannot proceed but, in this case the operator will keep reconciling in the background waiting for the problem to be fixed